### PR TITLE
Call dtor for all VarDeclarations with needsScopeDtor()==true.

### PR DESCRIPTION
**Pull request by JohanEngelen**
Fixes issue #4427

### CODE REVIEW

**Comment by JohanEngelen** [comment]
This PR brings us on-par with DMD and GDC, but there is more bad news: the simple test case works with DMD and GDC, but slightly more complex and it fails for all https://issues.dlang.org/show_bug.cgi?id=24010.  The destructors are called at the end of their definition, not at the end of their scope.

File: gen/llvmhelpers.cpp [hunk 1]
```diff
@@ -954,6 +963,11 @@ void DtoVarDeclaration(VarDeclaration *vd) {
       toElem(ex->exp);
     }
   }
+
+  if (!vd->isDataseg() && vd->needsScopeDtor()) {
```
Comment (kinke) [inline gen/llvmhelpers.cpp:967]: (There's an assertion at the top wrt. `!vd->isDataseg()`)

File: gen/llvmhelpers.cpp [hunk 2]
```diff
@@ -876,6 +876,15 @@ void DtoResolveVariable(VarDeclaration *vd) {
  * DECLARATION EXP HELPER
  ******************************************************************************/
 
+void pushVarDtorCleanup(IRState *p, VarDeclaration *vd) {
```
Comment (kinke) [inline gen/llvmhelpers.cpp:879]: (This might be nicer as `IRState` method)

**Review by kinke — APPROVED** [review]

**Comment by kinke** [comment]
@JohanEngelen: Are you unsure about this (and including it in v1.33)? Just wondering why it isn't merged yet (I was on vacation...).

**Comment by JohanEngelen** [comment]
I was slightly unsure about the implementation, but also that the bug would introduce another bug.
This is now properly fixed upstream: https://github.com/dlang/dmd/pull/15351 
Tuples should now no longer reach this code (if I understand the fix correctly).

**Comment by JohanEngelen** [comment]
@kinke This can be a very disruptive fix. So let's postpone until next release, when also DMD frontend has the additional fix https://github.com/dlang/dmd/pull/15351 . That way, we don't change user code behavior into a new buggy behavior, but instead wait until we have the full proper fix.

**Comment by JohanEngelen** [comment]
I believe this is fixed in 1.35.0 due to upstream fix.

**Pull request closed without merging**